### PR TITLE
Ensure encoding["source"] is available for a pathlib.Path object

### LINKED ISSUE #5888: open_[mf]dataset ds.encoding['source'] for pathlib.Path?
The question: do you want to support pathlib objects arriving at ds.encoding['source']?

**What happened**:
If you pass a pathlib object to open_[mf]dataset, the ds.encoding['source'] is not set.

**What you expected to happen**:
I believe that Paths used to work and now they dont, which greatly confused me. 

**Minimal Complete Verifiable Example**:

```python
(hv) jamesmcc@panama[523]:~> cat enc_source_pathlib.py
import numpy as np
from pathlib import Path
import xarray as xr

# inspiration in the test
# https://github.com/pydata/xarray/blob/97887fd9bbfb2be58b491155c6bb08498ce294ca/xarray/tests/test_backends.py#L4918

rnddata = np.random.randn(10)
ds = xr.Dataset({"foo": ("x", rnddata)})
file_name = "rnddata.nc"
ds.to_netcdf(file_name)

with xr.open_dataset(file_name) as ds:
    assert ds.encoding["source"] == file_name

with xr.open_dataset(Path(file_name)) as ds:
    print(ds.encoding)
    assert ds.encoding["source"] == file_name



(hv) jamesmcc@panama[524]:~/> ipython enc_source_pathlib.py
/Users/jamesmcc/python_libs/xarray/xarray/backends/cfgrib_.py:28: UserWarning: Failed to load cfgrib - most likely there is a problem accessing the ecCodes library. Try `import cfgrib` to get the full error message
  "Failed to load cfgrib - most likely there is a problem accessing the ecCodes library. "
/Users/jamesmcc/python_libs/xarray/xarray/backends/plugins.py:61: RuntimeWarning: Engine 'cfgrib' loading failed:
Cannot find the ecCodes library
  warnings.warn(f"Engine {name!r} loading failed:\n{ex}", RuntimeWarning)
{'unlimited_dims': set()}
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/python_libs/xarray/enc_source_pathlib.py in <module>
     16 with xr.open_dataset(Path(file_name)) as ds:
     17     print(ds.encoding)
---> 18     assert ds.encoding["source"] == file_name
     19 
     20 

KeyError: 'source'
```

I scanned git blame, but didnt turn up any obviously undesired changes.

### PR DESCRIPTION
**Pull request by ColemanTom**
Closes #5888. At the moment if you pass a `Path` object, the `source` will not be encoded in, only if a string is passed in. Given there is already a function to handle this, `_normalize_path`, I've simplified the logic to just run that over the `str/Path` to ensure it is always encoded.

- [x] Closes #5888
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

I'm not sure if this needs including in the `whats-new.rst`. I'm happy to add it if desired.

### CODE REVIEW

**Review by mathause — COMMENTED** [review]
I think it's also possible to pass the whole dataset as bytes & we don't want this to end off in `ds.encoding["source"]` I hope my suggestion would take care of this.

**Comment by ColemanTom** [comment]
> I think it's also possible to pass the whole dataset as bytes & we don't want this to end off in `ds.encoding["source"]` I hope my suggestion would take care of this.

Thanks. I had forgotten you could, for example, pass file-like objects as I don't tend to. I've confirmed your suggestion works and have added it to the change.

File: doc/whats-new.rst [hunk 1]
```diff
@@ -59,6 +59,8 @@ Bug fixes
   By `András Gunyhó <https://github.com/mgunyho>`_.
 - Avoid use of random numbers in `test_weighted.test_weighted_operations_nonequal_coords` (:issue:`6504`, :pull:`6961`).
   By `Luke Conibear <https://github.com/lukeconibear>`_.
+- :py:attr:`DataSet.encoding['source']` now exists when it was created from a Path object instead of only existing when a string path is provided (:issue:`5888`, :pull:`6974)
```
Comment (dcherian) [inline doc/whats-new.rst:62]:
```suggestion
- ``Dataset.encoding['source']`` now exists when reading from a Path object (:issue:`5888`, :pull:`6974)
```

File: doc/whats-new.rst [hunk 2]
```diff
@@ -65,6 +65,8 @@ Bug fixes
   By `András Gunyhó <https://github.com/mgunyho>`_.
 - Avoid use of random numbers in `test_weighted.test_weighted_operations_nonequal_coords` (:issue:`6504`, :pull:`6961`).
   By `Luke Conibear <https://github.com/lukeconibear>`_.
+- ``Dataset.encoding['source']`` now exists when reading from a Path object (:issue:`5888`, :pull:`6974)
+  By `Thomas Coleman <https://github.com/ColemanTom`_.
```
Comment (mathause) [inline doc/whats-new.rst:69]:
```suggestion
  By `Thomas Coleman <https://github.com/ColemanTom>`_.
```

**Comment by mathause** [comment]
Let's see if that works now - should auto-merge on green.

Thanks for your contribution and welcome to xarray!

File: doc/whats-new.rst [hunk 3]
```diff
@@ -65,6 +65,8 @@ Bug fixes
   By `András Gunyhó <https://github.com/mgunyho>`_.
 - Avoid use of random numbers in `test_weighted.test_weighted_operations_nonequal_coords` (:issue:`6504`, :pull:`6961`).
   By `Luke Conibear <https://github.com/lukeconibear>`_.
+- ``Dataset.encoding['source']`` now exists when reading from a Path object (:issue:`5888`, :pull:`6974)
```
Comment (mathause) [inline doc/whats-new.rst:68]:
```suggestion
- ``Dataset.encoding['source']`` now exists when reading from a Path object (:issue:`5888`, :pull:`6974`)
```

**Comment by kwodzicki** [comment]
It would be great to further improve this by handling file-like objects. I am running into an issue when trying to read some data from S3 buckets. I can pass in a file-like object and read the data, but there is no `.encoding['source']` in the subsequent dataset.

Can simulate this 'issue' using local data:
>     import xarray as xr
>     file = '/path/to/file.nc'
>     fid  = open(file, 'rb')
>     ds   = xr.open_dataset( fid, engine='h5netcdf' )
This will have no source encoding in the dataset, but it could be grabbed from `fid.name`.

Could a check be added for file-like objects, and then insert the `.path` or `.name` attribute into `.encoding['source']`?

**Pull request merged**